File: backend/src/main/java/com/whiteblue/backend/security/OAuth/OAuthUser.java
package com.whiteblue.backend.security.OAuth;

import com.whiteblue.backend.domain.user.User;
import lombok.Getter;
import org.springframework.security.core.GrantedAuthority;
import org.springframework.security.core.authority.SimpleGrantedAuthority;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.security.oauth2.core.user.OAuth2User;

import java.util.Collection;
import java.util.Collections;
import java.util.Map;

@Getter
public class OAuthUser implements UserDetails, OAuth2User {
    private final String username;

    private final String name;

    private final Collection<? extends GrantedAuthority> authorities;

    private final Map<String, Object> attributes;

    public OAuthUser(User user, Map<String, Object> attributes) {
        this.username = user.getUsername();
        this.name = user.getName();
        this.authorities = Collections.singletonList(new SimpleGrantedAuthority(user.getRole()
                                                                                    .getName()));
        this.attributes = attributes;
    }

    @Deprecated
    @Override
    public String getPassword() {
        return null;
    }

    @Override
    public boolean isAccountNonExpired() {
        return true;
    }

    @Override
    public boolean isAccountNonLocked() {
        return true;
    }

    @Override
    public boolean isCredentialsNonExpired() {
        return true;
    }

    @Override
    public boolean isEnabled() {
        return true;
    }

}

File: backend/src/main/kotlin/com/whiteblue/backend/security/oAuth/OAuthUser.kt
package com.whiteblue.backend.security.oAuth

import jakarta.validation.constraints.Email
import org.springframework.security.core.GrantedAuthority
import org.springframework.security.core.userdetails.UserDetails
import org.springframework.security.oauth2.core.user.OAuth2User

data class OAuthUser(
    val id: Long,

    @field:Email
    private val username: String,

    private val name: String,

    private val authorities: List<GrantedAuthority>,

    private val attributes: Map<String, Any>?
): UserDetails, OAuth2User {
    override fun getName(): String {
        return name
    }

    override fun getPassword(): String {
        TODO("Not yet implemented")
    }

    override fun getUsername(): String {
        return username
    }

    override fun isAccountNonExpired(): Boolean {
        TODO("Not yet implemented")
    }

    override fun isAccountNonLocked(): Boolean {
        TODO("Not yet implemented")
    }

    override fun isCredentialsNonExpired(): Boolean {
        TODO("Not yet implemented")
    }

    override fun isEnabled(): Boolean {
        TODO("Not yet implemented")
    }

    override fun getAttributes(): Map<String, Any>? {
        return attributes
    }

    override fun getAuthorities(): List<GrantedAuthority> {
        return authorities
    }
}
